ipa-client-samba utility

abbra opened the pull request.

rcritten commented:
Not a detailed review but moving a bunch of functions breaks the python API for other users. @tiran  handled this with a wrapper that emitted a DeprecationWarning in d5d9233b7c94eb33d6c7341899543224c14d2287

abbra commented:
@rcritten I can add back wrappers. I changed internal use, not sure if we ever had any external ones for these helpers as they are inside server code. Do you use them in the healthcheck tool?

rcritten commented:
I haven't looked in detail yet but this set off my alarm bells :-) A fair bit of the *instance and installutils functions/methods are called though.

tiran commented on client/ipa-client-samba.in:
```diff
@@ -0,0 +1,617 @@
+            statestore)
+
+def uninstall(fstore, statestore, options):
+    global host_princ
```
Please try to avoid globals. Amongst others, mutable globals make it harder to follow the code and complicate @t-woerner Ansible work. 

Is there any reason you cannot pass the vars around?

tiran commented on daemons/ipa-kdb/ipa_kdb_mspac.c:
```diff
@@ -417,11 +419,21 @@ static krb5_error_code ipadb_fill_info3(struct ipadb_context *ipactx,
     free(objectclasses);
 
+    if (is_idobject && is_ipauser) {
+	is_user = true;
```
Mixed spaces / tabs?

tiran commented on daemons/ipa-sam/ipa_sam.c:
```diff
@@ -3370,13 +3375,45 @@ static bool init_sam_from_ldap(struct ipasam_private *ipasam_state,
 
-	if (!(username = smbldap_talloc_first_attribute(priv2ld(ipasam_state),
-					entry, LDAP_ATTRIBUTE_UID, tmp_ctx))) {
+	usernames = ldap_get_values_len(priv2ld(ipasam_state), entry,
```
Mixed spaces / tabs?

tiran commented on ipaserver/install/plugins/adtrust.py:
```diff
@@ -692,3 +694,97 @@ def execute(self, **options):
+@register()
+class update_host_cifs_keytabs(Updater):
+    """
+    Synchronize host keytab and Samba keytab
```
Minor nit pick: The title line belongs on the same line as the triple quotes.

tiran commented on ipatests/test_integration/test_smb.py:
```diff
@@ -0,0 +1,158 @@
+        tasks.uninstall_master(cls.master)
+
+    def test_prepare_users(self):
+        global users
```
You don't need ``global`` here.

tiran commented on ipatests/test_integration/test_smb.py:
```diff
@@ -0,0 +1,158 @@
+           Test checks that both kernel SMB3 driver and userspace
+           smbclient utility work against IPA-enrolled Samba server
+        """
+        global users
```
You don't need ``global`` here.

tiran commented on ipatests/test_integration/test_smb.py:
```diff
@@ -0,0 +1,158 @@
+            smbclt.run_command(['kdestroy', '-A'])
+
+    def test_uninstall_samba(self):
+        global users
```
You don't need ``global`` here.

abbra commented:
@tiran I've updated the code to follow your recommendations except the mixed tabs issue in ipasam module. I'm not going to change that one -- I tried refactoring it in past and bringing it to a single type (tabs or spaces) means killing the history. I used tabs as that function is using tabs already.

I moved code from `client/ipa-client-samba.in` to `ipaclient/install/ipa_client_samba.py`, similar to existing `client/ipa-client-install.in`.

I removed the use of globals.

abbra commented:
@tiran, @wladich, @rcritten I've updated the design docs. I need to add more to the document about changes on the domain controller side but I think we can already start reviewing the client side:
* Domain member design doc: https://github.com/abbra/freeipa/blob/ipa-client-samba/doc/designs/adtrust/samba-domain-member.md
* Domain controller design doc:" https://github.com/abbra/freeipa/blob/ipa-client-samba/doc/designs/adtrust/samba-domain-controller.md

abbra commented:
I submitted documentation update that I consider final for this PR.

cryptomilk commented on doc/designs/adtrust/samba-domain-member.md:
```diff
@@ -0,0 +1,429 @@
+is handled by a separate `smbd` child;
+
+- `winbindd`, a process to perform identity resolution for all configured and
+  known domains. Each domain is handled by a separate `winbindd` child.
```
Only domains which have an active connection are handled by a child in the meantime. This means you have only a child for your own domain running. There could be another child for a trusted domain if idmap_ad is configured. Normally all trusted domain users auth's are sent to your own domain.

cryptomilk commented on doc/designs/adtrust/samba-domain-member.md:
```diff
@@ -0,0 +1,429 @@
+principals of the same host. In Active Directory all Kerberos principals
+associated with the host (service principal names, SPNs) share the same Kerberos
+keys. Thus, Samba needs to known a clear text password for the machine account
+and it can derive all Kerberos keys for itself based on that knowledge.
```
And to initiate a shared secret rotation (change the password).

rcritten commented on daemons/ipa-sam/ipa_sam.c:
```diff
@@ -3554,6 +3554,77 @@ static NTSTATUS getsam_interdom_trust_account(struct pdb_methods *methods,
+
+	sid_str = sid_talloc_string(ipasam_state->idmap_ctx, tmp_ctx, sid);
+	if (sid_str == NULL) {
+		return NT_STATUS_NO_MEMORY;
```
This will leak tmp_ctx. It should set status and goto done so that talloc_free(tmp_ctx) can be called.

abbra replied:
Noted.

rcritten commented on daemons/ipa-sam/ipa_sam.c:
```diff
@@ -3554,6 +3554,77 @@ static NTSTATUS getsam_interdom_trust_account(struct pdb_methods *methods,
+			     &result);
+	if (ret != LDAP_SUCCESS) {
+		status = NT_STATUS_NO_SUCH_USER;
+		goto done;
```
Is this safe? What about search timeouts, ACI and other sorts of problems?

abbra replied:
We have the same code in `ipasam_getsampwnam()` and other places for years.

rcritten replied:
What caught my eye here was another ticket which I can't find now but it was related to IPA incorrectly returning NO_SUCH_USER (or notfound) in a 389-ds plugin when it was I think related to a timeout. IIRC there was a bit of back and forth before Sumit finally discovered the issue.

abbra replied:
Ah, no that issue was totally different. That was extop plugin action which is triggered by the use of extended operation by SSSD. It is not the same as we have here. I don't think it is worth the check here. If `cifs/..` principal loses ACIs, we'll notice it earlier because we rely on a number of other ACIs here assigned to the trust agent group.

rcritten commented on daemons/ipa-sam/ipa_sam.c:
```diff
@@ -3554,6 +3554,77 @@ static NTSTATUS getsam_interdom_trust_account(struct pdb_methods *methods,
+
+	count = ldap_count_entries(priv2ld(ipasam_state), result);
+	if (count != 1) {
+		status = NT_STATUS_NO_SUCH_USER;
```
Perhaps log something? This could be difficult to debug.

abbra replied:
I added, but it is the same for other methods already for quite some time.

rcritten commented on daemons/ipa-kdb/ipa_kdb_mspac.c:
```diff
@@ -417,11 +419,21 @@ static krb5_error_code ipadb_fill_info3(struct ipadb_context *ipactx,
     }
     free(objectclasses);
 
+    if (is_idobject && is_ipauser) {
```
Can you add a block comment with a message similar to commit message so the reason for this is clear?

abbra replied:
Done.

rcritten replied:
BTW my inspiration for suggesting adding it to the code was the great commit message overall! They all really helped me get a handle on what each patch was doing.

rcritten commented on ipaserver/install/plugins/adtrust.py:
```diff
@@ -716,3 +718,96 @@ def execute(self, **options):
+            if not (copied or uptodate):
+                self.copy_key(paths.SAMBA_KEYTAB, hostkey)
+
+        return False, []
```
Does this mean if someone gets a new host keytab they will need to run this upgrade code in order to update the cifs principal? If someone does this would there be reasonably understandable error messages or just access denied?

abbra replied:
They can re-run ipa-adtrust-install which will copy the keytab over. I think the error message in Samba logs will show the problem  as `GSS VerifyMic failed: ...`:
(from `samba/auth/kerberos/gssapi_helper.c:NTSTATUS gssapi_check_packet()`)
```
        if (GSS_ERROR(maj_stat)) {
                char *error_string = gssapi_error_string(NULL,
                                                         maj_stat,
                                                         min_stat,
                                                         mech);
                DEBUG(1, ("GSS VerifyMic failed: %s\n", error_string));
                talloc_free(error_string);

                return NT_STATUS_ACCESS_DENIED;
        }
```

rcritten replied:
Ok. This may lead us to verifying the two keytabs are consistent. If not we can always add something later.

abbra replied:
The verification is running every time we upgrade data. So any package update, for example. If not, re-running `ipa-adtrust-install` will do this for you too. Finally, making an update file `90-update-host-cifs-keytabs.update` with
```
plugin: update_host_cifs_keytabs
```
and running it with `ipa-ldap-updater ./90-update-host-cifs-keytabs.update` would do the same action.

rcritten commented on ipaserver/plugins/service.py:
```diff
@@ -447,6 +447,16 @@ class service(LDAPObject):
             ],
             'default_privileges': {'Service Administrators'},
         },
+        'System: Read POSIX details of the services': {
```
Could I suggest instead "Read POSIX details of SMB services"?

rcritten commented on client/man/ipa-client-samba.1:
```diff
@@ -0,0 +1,88 @@
+
+5. Populate Samba databases by setting the domain details and the randomly generated machine account password from the previous step.
+
+6. Create a default [homes] share to allow users to log in to their home directories if \-\-no\-mkhomedir option was not specified.
```
I think it would read better as:

...to their home directories unless the \-\-no\-homes option was specified.

Either way the option name needs to be fixed.

rcritten commented on freeipa.spec.in:
```diff
@@ -520,6 +520,22 @@ If your network uses IPA for authentication, this package should be
+
+%description client-samba
+This package provides command-line tools to deploy Samba domain member
+on the machine enrolled into a FreeIPA environment
```
Is this packaged separately to enforce the idea that other work is needed to setup a Samba share (setup Trust)? Is there any other reason not to just include this alongside ipa-client-install?

abbra replied:
If you would have it in freeipa-client package, then it would pull in Samba packages on every deployment. I don't think this is really intended.

rcritten replied:
Heh, no, I didn't read this carefully enough! A separate package it is :-)

rcritten commented on ipaclient/install/ipa_client_samba.py:
```diff
@@ -0,0 +1,726 @@
+
+def pretty_print_domain_information(info):
+    result = []
+    for d in info:
```
maybe domain instead of d? Or is it obvious?

abbra replied:
Turned to domain

rcritten commented on ipaclient/install/ipa_client_samba.py:
```diff
@@ -0,0 +1,726 @@
+        domains = retrieve_domain_information(api)
+        str_info = pretty_print_domain_information(domains)
+        logger.info("Discovered domains to use:\n%s", str_info)
+        print("Discovered domains to use:\n%s" % str_info)
```
What if no domains are returned? Should this be allowed to continue?

abbra replied:
At least our own domain should be discovered. I added the code that checks for two parts:
- `ipa trustconfig-show` actually available through the metadata
- `ipa trustconfig-show` returns at least information about our domain

rcritten commented on ipaclient/install/ipa_client_samba.py:
```diff
@@ -0,0 +1,726 @@
+        dest="netbiosname",
+        help="NetBIOS name of the machine",
+        default=None,
+    )
```
It doesn't look to me that the server option determines which server the client connects to.

rcritten commented on ipaclient/install/ipa_client_samba.py:
```diff
@@ -0,0 +1,726 @@
+        dest="force",
+        action="store_true",
+        default=False,
+        help="force installation by redoing already done steps",
```
force installation by redoing all steps

rcritten commented on ipaclient/install/ipa_client_samba.py:
```diff
@@ -0,0 +1,726 @@
+        dest="debug",
+        action="store_true",
+        default=False,
+        help="enable debugging",
```
We typically say "print debugging information"

rcritten commented on ipaclient/install/ipa_client_samba.py:
```diff
@@ -0,0 +1,726 @@
+    # by the code called from 'net changesecretpw -f'. Thus, we'd limit
+    # password to ASCII only.
+    password = generate_random_password(128, 255)
+    return password
```
return generate_random_password(128, 255)

rcritten commented on ipaclient/install/ipa_client_samba.py:
```diff
@@ -0,0 +1,726 @@
+    except ipautil.CalledProcessError as e:
+        logger.error(
+            "Cannot set machine account password at IPA DC. " "Error: %s",
+            str(e),
```
I don't believe str is needed because it will be converted in %s.

rcritten commented on ipaclient/install/ipa_client_samba.py:
```diff
@@ -0,0 +1,726 @@
+    for svc in (smb, winbind):
+        if svc.is_running():
+            svc.stop()
+        svc.disable()
```
Should we disable services we don't enable?

abbra replied:
We should because we removed the configuration.

abbra commented on ipapython/ipautil.py:
```diff
@@ -391,7 +391,8 @@ def __str__(self):
 def run(args, stdin=None, raiseonerr=True, nolog=(), env=None,
```
I need output returned but not logged.

rcritten replied:
Ah ok, understood.

abbra commented on daemons/ipa-sam/ipa_sam.c:
```diff
@@ -3554,6 +3554,77 @@ static NTSTATUS getsam_interdom_trust_account(struct pdb_methods *methods,
 	return status;
 }
 
```
Done

abbra commented:
I think I addressed all review comments and tests have passed.

rcritten commented:
ipa-client-samba correctly fails to run on master
ipa-client-samba fails with a double error message without a trust:

```
Searching for IPA server...
IPA server: DNS discovery
Chosen IPA master: ipa.example.test
SMB principal to be created: cifs/ipa.example.test@EXAMPLE.TEST
NetBIOS name to be used: IPA
No configured trust controller detected on IPA masters. Use ipa-adtrust-install on an IPA master to configure trust controller role.
No configured trust controller detected on IPA masters. Use ipa-adtrust-install on an IPA master to configure trust controller role.
```
But correctly had return code of 1.

abbra commented:
@rcritten the double printing is due to logger.error() and print() -- in ipalib/pluggable.py:IPA.bootstrap() we do install additional handler that prints errors on stderr. If you want, I can remove the explicit print()

rcritten commented on ipaserver/plugins/service.py:
```diff
@@ -664,6 +674,133 @@ def post_callback(self, ldap, dn, entry_attrs, *keys, **options):
+        entry_attrs['cn'] = netbiosname
+        entry_attrs['homeDirectory'] = '/dev/null'
+        entry_attrs['uidNumber'] = '-1'
+        entry_attrs['gidNumber'] = '-1'
```
Please change both -1 to baseldap.DNA_MAGIC

abbra replied:
Done.

abbra commented:
master:

* 84201e1daff7e2cd80352654dc7c81530b51dc74 adtrust: add design document for Samba domain member on IPA client
* cdb94e0ff2c7bc03b2f0064b77fedabfa0ae8121 ipaserver.install.installutils: move commonly used utils to ipapython.ipautil
* d85e0550cab58d64e65610314906d32bd21a9e39 ipapython.ipautil.run: allow skipping stdout/stderr logging
* a42352628df6ac4a9a174bee89032c356fac4801 ipasam: add lookup of an account by SID
* 91abd1f67ad86ecdaf3100e282b045a6be7de591 ipasam: add handling of machine accounts
* 653f72079ee27cfdfb880afb20905b7099748c14 kdb: support SMB services on IPA domain members
* d631e008ccec6827ca0c4b8d442469b0f0c994a8 adtrust: update Samba domain controller keytab with host keys
* afb8305ada944293f978a20f3829d0ab93180c2d ipaserver.plugins.service: add service-add-smb to set up an SMB service
* 814592cf2218956893baa2272101fffa93abb465 ipa-client-samba: a tool to configure Samba domain member on IPA client
* e25392e976bcbc34e53fafa5eb68014ff3ceb5e5 prci: add test_integration/test_smb to the gating set
